### PR TITLE
Implement #3

### PR DESCRIPTION
**Pull request by TheEpicBlock**
This implements the test vectors as outlined in #3 for all implementations except for dlang. I copied over all testcases I could find. All testcases succeed except for `a-a < a`, which was included in go's testcases but fails in all other implementations, I'm not sure what the correct behaviour is here.

Instead of the large int test being optional, it's in a separate file and implementations chose whether to include that file in their tests. This means that the large int test is fully enforced for those languages (so if it regresses you would notice).

- [x] I own the full rights to my contribution, and agree to release it under the terms of the Creative Commons Zero Public Domain Dedication

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -161,70 +161,81 @@ impl Ord for FlexVer<'_> {
...
+    fn test_clamp() {
         assert_eq!(
-            compare(&right, &left),
-            match result {
-                Less => Greater,
-                Greater => Less,
-                Equal => Equal,
-            }
+            FlexVer("1.1.0"),
+            FlexVer("1.1.0").clamp(FlexVer("1.0.0"), FlexVer("1.2.0"))
```
Comment (Infinidoge) [inline rust/src/lib.rs]: Nitpick: I think `test_clamp` should be below `test_min`/`test_max`
It just feels like the right place.

File: js/test.js [hunk 2]
```diff
@@ -1,43 +1,43 @@
...
+				
+				var lefthand = split[0];
+				var righthand = split[2];
+				var ordening = split[1];
 
-test("1.0", "1.1", -1);
+				return runTest(lefthand, righthand, ordening)
+			})
+	})
+})
```
Comment (Infinidoge) [inline js/test.js]:
```suggestion
})

```
Nitpick: Should have a newline at the end of the file

File: java/src/test/java/com/unascribed/flexver/Ordering.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+     * Converts an integer returned by a method like {@link FlexVerComparator#compare(String, String)} to an {@link Ordering}
+     */
+    public static Ordering fromComparison(int i) {
+        if (i < 0) return Ordering.Less;
+        if (i == 0) return Ordering.Equal;
+        if (i > 0) return Ordering.Greater;
+
+        throw new IllegalStateException();
+    }
+}
```
Comment (Infinidoge) [inline java/src/test/java/com/unascribed/flexver/Ordering.java]:
```suggestion
}

```
Nitpick: Should have a newline at the end of the file

File: java/src/test/java/com/unascribed/flexver/Ordering.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package com.unascribed.flexver;
+
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
+public enum Ordering {
+    Less("<"),
+    Equal("="),
+    Greater(">");
```
Comment (exaskye) [inline java/src/test/java/com/unascribed/flexver/Ordering.java]: Conventionally, enum constants (all constants, in fact) are SCREAMING_SNAKE_CASE.

```suggestion
    LESS("<"),
    EQUAL("="),
    GREATER(">");
```
(Suggestion must be applied via IDE refactor, not in GitHub web.)

File: java/src/test/java/com/unascribed/flexver/Ordering.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Converts an integer returned by a method like {@link FlexVerComparator#compare(String, String)} to an {@link Ordering}
+     */
+    public static Ordering fromComparison(int i) {
+        if (i < 0) return Ordering.Less;
+        if (i == 0) return Ordering.Equal;
+        if (i > 0) return Ordering.Greater;
+
+        throw new IllegalStateException();
```
Comment (exaskye) [inline java/src/test/java/com/unascribed/flexver/Ordering.java]: Should be `AssertionError`

File: java/src/test/java/com/unascribed/flexver/Ordering.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            case Equal:
+                return this;
+            case Greater:
+                return Ordering.Less;
+            default:
+                throw new IllegalArgumentException();
+        }
+    }
+
+    public static Ordering fromStr(String str) {
```
Comment (exaskye) [inline java/src/test/java/com/unascribed/flexver/Ordering.java]: from or fromString

File: test/test_vectors.txt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+# Dash is included in prerelease comparison (if stripped it will be a smaller component)
+# Note that a-a < a=a regardless since the prerelease splits the component creating a smaller first component; 0 is added to force splitting regardless
+a0-a < a0=a
+
+# Pre-releases must contain only non-digit
+1.16.5-10 > 1.16.5
+
+# Pre-releases can have multiple dashes (should not be split)
+# Reasoning for test data: "p-a!" > "p-a-" (correct); "p-a!" < "p-a t-" (what happens if every dash creates a new component)
+-a- > -a!
```
Comment (unascribed) [inline test/test_vectors.txt:41]: Didn't realize the Go impl had some new good tests in it like this one.

File: test/test_vectors.txt [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+1.5 < 1.5-2
+1.5-pre10 > 1.5-pre2
+
+# Empty strings
+ = 
+1 > 
+ < 1
+
+# Check boundary between textual and prerelease
+a-a < a
```
Comment (exaskye) [inline test/test_vectors.txt:27]: I believe this describes the correct behavior. Why are all impls but Go failing this...?

File: java/src/test/java/com/unascribed/flexver/FlexVerBasicTest.java [hunk 9]
```diff
@@ -1,92 +1,49 @@
...
-		if (c == 0) res = "=";
-		if (c > 0) res = ">";
-		System.out.println(represent(a)+(ANSI?"\u001B[0m ":" ")+res+" "+represent(b));
-	}
-	
-	private static int signum(int i) {
-		return i < 0 ? -1 : i > 0 ? 1 : 0;
-	}
+	@ParameterizedTest(name = "{0} {2} {1}")
+	@MethodSource("getEqualityTests")
```
Comment (unascribed) [inline java/src/test/java/com/unascribed/flexver/FlexVerBasicTest.java:21]: Well, I learned something new about JUnit today.

File: java/build.gradle [hunk 10]
```diff
@@ -8,6 +8,20 @@ sourceCompatibility = targetCompatibility = 8
...
 version = '1.0.2'
 
+repositories {
+	mavenCentral()
+}
+
+dependencies {
+    testImplementation 'org.junit.jupiter:junit-jupiter-api:5.9.2'
+    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.9.2'
+	implementation 'org.junit.jupiter:junit-jupiter-params:5.9.2'
```
Comment (exaskye) [inline java/build.gradle]: This will pull in this dep transitively on Maven! Needs to be `compileOnly` or `testImplementation`.

**Review by exaskye — COMMENTED** [review]
Oh, I wasn't expecting someone to PR this; that was a TODO for myself, heh. Thanks!

I can only satisfactorily review the Java and JS implementations; going to need codeowner reviews for the others.

JS: :heavy_check_mark: 
Java: :heavy_minus_sign: (but just nits)

**Review by Infinidoge — APPROVED** [review]
The Rust implementation looks good to me!
(Though I'm not sure about the `a-a < a` situation either, will need Una's input on that)

Had a few nitpicks, mainly about newlines at the ends of files, but other that the PR looks solid.

I've only personally tested the Rust implementation.

File: test/test_vectors.txt (hunk 8 above)
Comment (exaskye) [reply test/test_vectors.txt:27]: @Infinidoge seems to have figured it out; in Go, `-` is treated specially and splits out a component, but all the other implementations look for *textual components* that are *already split off* that start with `-`.

Checking the spec, that *is* correct behavior, but I don't know if that's the *intended* behavior...

File: java/src/test/java/com/unascribed/flexver/Ordering.java (hunk 4 above)
Comment (TheEpicBlock) [reply java/src/test/java/com/unascribed/flexver/Ordering.java]: That's pretty embarrassing. Guess I've been doing too much Rust lately

File: go/flexver/main_test.go [hunk 11]
```diff
@@ -9,219 +9,181 @@
...
-func opChar(op int) string {
-	if op < 0 {
-		return "<"
-	} else if op == opEQ {
-		return "="
-	} else if op == opGT {
-		return ">"
-	} else {
-		return "!"
+func RunCompare(t *testing.T, lefthand string, righthand string, ordening int) {
```
Comment (comp500) [inline go/flexver/main_test.go]: Spelling mistake in `ordening` (also for `RunLess`/`RunEqual`)

File: go/flexver/main_test.go [hunk 12]
```diff
@@ -9,219 +9,181 @@
...
-				// Catch panic
-				defer func() { _ = recover() }()
-			}
+func TestStandardized(t *testing.T) {
+	for _, test := range ENABLED_TESTS {
+		file, err := os.Open("../../test/" + test)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer file.Close()
```
Comment (comp500) [inline go/flexver/main_test.go]: This won't close each individual test data file until the whole function returns; the for loop contents could be refactored out into another function (or inline closure)

File: go/flexver/main_test.go [hunk 13]
```diff
@@ -9,219 +9,181 @@
...
-					(v.op == opGT && res == true) ||
-					(v.op == shouldError) {
-					t.Fatalf("Unexpected result %v", res)
-				}
-			} else if v.op != shouldError {
-				t.Fatalf("Unexpected error %v", err)
-			}
-		})
+func TestInvalid(t *testing.T) {
+	_, err := CompareError("\xc3\x28", "")
```
Comment (comp500) [inline go/flexver/main_test.go:157]: This no longer tests whether the panicking variants panic on invalid input; `defer func() { _ = recover() }()` can be used to catch a panic as done in the previous implementation.

**Review by comp500 — CHANGES_REQUESTED** [review]
The Go implementation looks good, just got a few suggestions.

**Review by comp500 — APPROVED** [review]

**Review by exaskye — APPROVED** [review]

**Pull request merged**